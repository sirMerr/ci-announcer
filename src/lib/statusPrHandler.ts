import { Context } from 'probot';
// import {LoggerWithTarget} from 'probot/lib/wrap-logger';
import { getTravisBuildData, getTravisErrorLogs } from './travis';

// interface EventInfo {
// 	repo: string,
// 	owner: string,
// 	log: LoggerWithTarget | null,
// 	sha: string
// }

// export let eventInfo: EventInfo = {
// 	repo: '',
// 	owner: '',
// 	log: null,
// 	sha: ''

// }

/**
 * Handler which will get the associated
 * Travis build info and prep a message
 * for the associated PR trigger
 *
 * @param context
 * @param owner
 * @param repo
 */
export const statusPrHandler = async ({
  context,
  owner,
  repo,
}: {
  context: Context;
  owner: string;
  repo: string;
}) => {
  const { log, payload, github } = context;
  const { sha } = payload;
  log.info('statusPrHandler');

  // Get the CI build data
  const buildData = await getTravisBuildData(context);
  const jobId = buildData.jobs[0].id;

  // Get the CI error logs formatted
  const errorLogs = await getTravisErrorLogs({
    context,
    jobId,
  });

  const body = makeIssueBody({
    errorLogs,
    sha,
    travisUrl: payload.target_url,
  });

  log.info(body);

  await github.issues.createComment({
    owner,
    repo,
    body,
    number: buildData.pull_request_number,
  });

  return;
};

const makeIssueBody = ({
  errorLogs,
  sha,
  travisUrl,
}: {
  errorLogs: string;
  sha: string;
  travisUrl: string;
}): string => {
  return (
    `The [TravisCI build](${travisUrl}) failed as of ${sha}\n` +
    '<pre><code>' +
    errorLogs +
    '</code></pre>'
  );
};
