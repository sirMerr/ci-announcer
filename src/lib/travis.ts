import { Context } from 'probot';
import { WebhookPayloadWithRepository } from 'probot/lib/context';
import fetch from 'node-fetch';
import * as http from 'http';

const travisAPI = {
  url: String(process.env.TRAVIS_API_URL),
  headers: {
    'Travis-API-Version': '3',
    'User-Agent': 'API Explorer',
    Authorization: `token ${process.env.TRAVIS_ACCESS_TOKEN}`,
  },
};

/**
 * Make a request to /build
 * @param context Helpers for the GitHub webhook
 */
export const getTravisBuildData = async (context: Context) => {
  const { log, payload } = context;
  const buildId = getTravisBuildId(payload);

  log.info('BuildId', buildId);

  if (!buildId) return;

  const { url, headers } = travisAPI;
  log.info('GET: ', `${url}/build/${buildId}`);

  const res = await fetch(`${url}/build/${buildId}`, { headers });
  const json = await res.json();

  return json;
};

/**
 * Get Build ID from context for API call to
 * Travis
 * @param context Helpers for the GitHub webhook
 */
const getTravisBuildId = (
  payload: WebhookPayloadWithRepository
): null | number => {
  const regex = /(?<=builds\/)[^?]*/;
  const build = payload.target_url.match(regex);

  return build === null ? null : build[0];
};

/**
 * Make request to /job/log which has the logs
 * @param jobId Travis Build Job ID
 */
export const getTravisJobLog = async (context: Context, jobId: number) => {
  const { headers } = travisAPI;

  const options = {
    host: 'api.travis-ci.org',
    path: `/job/${jobId}/log`,
    headers,
  };
  let errorLog = '';

  http
    .get(options, function(resp) {
      context.log.info(resp);
      context.log.info('hi');
      let respContent = '';
      resp.on('data', function(data) {
        context.log.info('data');
        context.log.info(data.toString());
        respContent += data.toString(); //data is a buffer instance
      });
      resp.on('end', function() {
        context.log.info('END');
        context.log.info(respContent);
        errorLog = respContent;
      });
    })
    .on('error', context.log.error);

  return errorLog;
  //   const res = await fetch(`${url}/job/${jobId}/log.txt`, { headers });
  //   const text = await res.text();
  //   context.log.info(text);
  //     const indexFailStart = text.lastIndexOf('[0m[7m[1m[31m FAIL');
  //     const indexFailEnd = text.lastIndexOf('[999D[K[1mTest Suites:');

  //     const errorLog = text.slice(indexFailStart, indexFailEnd);

  //   return errorLog;
};
