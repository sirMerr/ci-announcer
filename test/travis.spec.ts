import fetch from 'node-fetch';
import { getTravisErrorLogs, getTravisBuildData } from '../src/lib/travis';
import {
  mockLogWithFailNotFinal,
  mockLogWithMultipleFails,
} from './fixtures/logs';

jest.mock('node-fetch');

describe('will run travis fetches and manipulation', () => {
  let context = {
    log: jest.fn(),
    payload: {
      target_url:
        'https://travis-ci.org/1fsdsa/ci-announcer-tester/builds/394661277?utm_source=github_status&utm_medium=notification',
    },
  } as any;

  context.log.info = jest.fn();
  context.log.error = jest.fn();

  it('should fetch travis build logs', async () => {
    // @ts-ignore
    fetch.mockImplementation(() => Promise.resolve({ json: () => 'dataMock' }));

    const buildData = await getTravisBuildData(context);

    expect(buildData).toMatchSnapshot();
  });

  it('should fetch logs with failure that has logs between it and return proper travis error logs', async () => {
    // @ts-ignore
    fetch.mockImplementation(() => Promise.resolve(mockLogWithFailNotFinal));

    const errorLogs = await getTravisErrorLogs({ context, jobId: 1 });

    expect(errorLogs).toMatchSnapshot();
  });

  it('should fetch logs with multiple fails and return proper travis error logs', async () => {
    // @ts-ignore
    fetch.mockImplementation(() => Promise.resolve(mockLogWithMultipleFails));

    const errorLogs = await getTravisErrorLogs({ context, jobId: 1 });

    expect(errorLogs).toMatchSnapshot();
  });

  // it('should fetch logs with no failing test and return proper error message', async () => {
  // 	// @ts-ignore
  // 	fetch.mockImplementation(() => Promise.resolve(mockLogWithNoFail));

  // 	console.log('dsdads');
  // 	const errorLogs = await getTravisErrorLogs({ context, jobId: 1 });

  // 	console.log(errorLogs);
  // 	expect(errorLogs).toBe('No logs found, please check your build above');
  // });
});
