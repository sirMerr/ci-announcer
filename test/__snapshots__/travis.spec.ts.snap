// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will run travis fetches and manipulation should fetch logs with failure that has logs between it and return proper travis error logs 1`] = `
" FAIL  ./fail.test.js
  ✕ mistakes 1 + 2 to equal 4 (13ms)

  ● mistakes 1 + 2 to equal 4

    expect(received).toBe(expected) // Object.is equality

    Expected: 4
    Received: 3

      1 | test('mistakes 1 + 2 to equal 4', () => {
    > 2 | 	expect(1 + 2).toBe(4);
        | 	              ^
      3 | });
      4 | 

      at Object.<anonymous>.test (fail.test.js:2:16)"
`;

exports[`will run travis fetches and manipulation should fetch logs with multiple fails and return proper travis error logs 1`] = `" FAIL b[2mfail.test.js:2:16)"`;

exports[`will run travis fetches and manipulation should fetch travis build logs 1`] = `"dataMock"`;
